src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,332 @@
+  }
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+template<typename OptimizerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Train(OptimizerType& Optimizer)
+{
+  if (!reset)
+    Reset();
+  offset = 0;
+  trainGenerator = false;
+  GenerateData();
+  Optimizer.Optimize(*this, parameter);
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+double GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Evaluate(const arma::mat& /*parameters*/,
+    const size_t i, const bool /*deterministic*/)
+{
+  currentInput = this->predictors.unsafe_col(i);
+  currentTarget = this->responses.unsafe_col(i);
+  Forward(std::move(currentInput));
+  double res = discriminator.outputLayer.Forward(std::move(
+      boost::apply_visitor(outputParameterVisitor,
+      discriminator.network.back())), std::move(currentTarget)); 
+  return res;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Gradient(const arma::mat& parameters, const size_t i, arma::mat& gradient)
+{
+  if (!reset)
+    Reset();
+
+  if (gradient.is_empty())
+  {
+    if (parameter.is_empty())
+    {
+      Reset();
+    }
+    gradient = arma::zeros<arma::mat>(parameter.n_rows, 1);
+  }
+  else
+  {
+    gradient.zeros();
+  }
+
+  // Gradient for generator network
+  gradientGenerator = arma::mat(gradient.memptr(),
+    generator.Parameters().n_rows, 1, false, false);
+
+  // Gradient for discriminator network
+  gradientDiscriminator = arma::mat(gradient.memptr() + gradientGenerator.n_elem,
+    discriminator.Parameters().n_rows, 1, false, false);
+
+  // get the gradients of the discriminator
+  discriminator.Gradient(parameters, i, gradientDiscriminator);
+  if (trainGenerator)
+  {
+    // pass the error from discriminator to generator
+    generator.error = boost::apply_visitor(deltaVisitor,
+        discriminator.network[1]);
+    generator.Backward();
+    generator.ResetGradients(gradientGenerator);
+    // set the current input requrired for gradient computation
+    generator.currentInput = predictors.col(i);
+    generator.Gradient();
+  }
+
+  // Freeze weights of discrminator if generator is training
+  if (!trainGenerator)
+    gradientGenerator.zeros();
+  else
+    gradientDiscriminator.zeros();
+
+  // if full batch seen then alternate training
+  if (iterations % numFunctions == 0)
+  {
+    if (!trainGenerator && iterationDiscriminator==1)
+    {
+      trainGenerator = true;
+      GenerateData();
+      iterationDiscriminator = 2 * disIteration;
+    }
+    else
+    {
+      trainGenerator = false;
+      offset = (offset + batchSize) % trainData.n_cols;
+      GenerateData();
+      iterationDiscriminator--;
+    }
+  }
+  iterations++;
+  std::cout << "gradient generator" << std::endl;
+  gradient.rows(0, 5).print();
```

Comment:
Have you checked the full gradient? We should make sure that each parameter is updated e.g. by setting the gradient to e.g. ``DBL_MAX`` and afterwards we could check if some of the gradient are still ``DBL_MAX``.